tools/clang/test/HLSL/out-param-diagnostics.hlsl

```diff
@@ -170,6 +170,20 @@ void MaybeUsedMaybeUnused([maybe_unused] out int Val, int Cnt) { // expected-not
     Val = 1;
 } // expected-note{{uninitialized use occurs here}}
 
+void Use(int V) {}
+
+void NoAnnotationIsUse(out int V) { // expected-note{{variable 'V' is declared here}}
+  Use(V); // expected-warning{{parameter 'V' is uninitialized when used here}}
+}
+
+RWByteAddressBuffer buffer;
+
+// No expected diagnostic here. InterlockedAdd is not annotated with HLSL
+// parameter annotations.
```

Comment:
This comment is confusing. Isn't the reason there's no diagnostic because it *is* annotated with parameter annotations?

Reply:
Confusingly and probably incorrectly it isn’t annotated. My comment lacks sufficient details here. This change ignores unannotated parameters and instead leaves it up to the rules for C/C++, which ignores values used as pass-by-reference parameters.

Reply:
Yeah, that is confusing.  I would think that the reason there's no diagnostic is that the third parameter *is* annotated with `out`, like an ordinary case where no warning should be emitted (and where we should have a test for that as you pointed out above).  Here's the intrinsic definition:
```
void [[]] InterlockedAdd64(in uint byteOffset, in u64 value, out any_int64 original) : interlockedadd_immediate;
```
Or is it the case that even though this has `out` in the intrinsic definition, it doesn't add the `HLSLOutAttr` to the parameter decl like an ordinary HLSL function would, yet we still need to test to ensure that it doesn't think this is a use of the uninitialized value?

Reply:
Yes, we don’t translate the annotations from the intrinsic definitions to the annotations in the AST, we just use them to generate the qual types for the arguments. `InterlockedAdd` ends up with a signature something like `InterlockedAdd(int, int, int && __restrict)`, with no HLSL annotations on the `ParamVarDecl`s

I don’t think many of these are generating correct AST representations, but since they’re completely missing the HLSL annotations the best thing we can do is rely on C/C++ rules.